pkg/console/service/service.go

```diff
@@ -0,0 +1,282 @@
+package service
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/emicklei/go-restful/v3"
+	authnv1 "k8s.io/api/authentication/v1"
+	core "k8s.io/api/core/v1"
+	rbac "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	kubevirt "kubevirt.io/api/core"
+	kubevirtv1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/client-go/kubecli"
+
+	"github.com/kubevirt/vm-console-proxy/api/v1alpha1"
+)
+
+const (
+	AppKubernetesNameLabel      = "app.kubernetes.io/name"
+	AppKubernetesPartOfLabel    = "app.kubernetes.io/part-of"
+	AppKubernetesVersionLabel   = "app.kubernetes.io/version"
+	AppKubernetesManagedByLabel = "app.kubernetes.io/managed-by"
+	AppKubernetesComponentLabel = "app.kubernetes.io/component"
+)
+
+type Service interface {
+	TokenHandler(request *restful.Request, response *restful.Response)
+	VncHandler(request *restful.Request, response *restful.Response)
+}
+
+func NewService(kubevirtClient kubecli.KubevirtClient) Service {
+	return &service{
+		kubevirtClient: kubevirtClient,
+	}
+}
+
+type service struct {
+	kubevirtClient kubecli.KubevirtClient
+}
+
+func (s *service) TokenHandler(request *restful.Request, response *restful.Response) {
+	err := ValidateVncRbac(request, s.kubevirtClient)
+	if err != nil {
+		_ = response.WriteError(http.StatusUnauthorized, err)
+		return
+	}
+
+	params, err := readTokenRequestParameters(request)
+	if err != nil {
+		_ = response.WriteError(http.StatusBadRequest, err)
+		return
+	}
+
+	vm, err := s.kubevirtClient.VirtualMachine(params.namespace).Get(params.name, &metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			_ = response.WriteError(http.StatusNotFound, fmt.Errorf("VirtualMachine does not exist: %w", err))
+			return
+		}
+		_ = response.WriteError(http.StatusInternalServerError, fmt.Errorf("error getting VirtualMachine: %w", err))
+		return
+	}
+
+	resourceName := vm.Name + "-vnc-access"
+	err = s.createResources(request.Request.Context(), resourceName, vm)
+	if err != nil {
+		_ = response.WriteError(http.StatusInternalServerError, fmt.Errorf("error creating resources: %w", err))
+		return
+	}
+
+	token, err := s.requestToken(request.Request.Context(), vm.Namespace, resourceName, params.duration)
+	if err != nil {
+		_ = response.WriteError(http.StatusInternalServerError, fmt.Errorf("failed to request token: %w", err))
+		return
+	}
+
+	_ = response.WriteAsJson(&v1alpha1.TokenResponse{
+		Token: token,
+	})
+}
+
+func (s *service) VncHandler(_ *restful.Request, response *restful.Response) {
```

Comment:
Do we need really this? Wouldn't implicitly returning 404 be enough?

Reply:
In case of upgrade, it is a good idea to indicate to user, that the endpoint was removed

Reply:
@akrejcir WDYT?